src/test/ui/libstd-case-typo/alloc.stderr

```diff
@@ -0,0 +1,25 @@
+error[E0412]: cannot find type `LayOut` in this scope
+  --> $DIR/alloc.rs:4:20
+   |
+LL | fn test_layout(_x: LayOut){}
+   |                    ^^^^^^ not found in this scope
+   |
+help: did you mean `std::alloc::Layout`?
+   |
+LL | fn test_layout(_x: std::alloc::Layout){}
+   |                    ^^^^^^^^^^^^^^^^^^
+
+error[E0412]: cannot find type `system` in this scope
+  --> $DIR/alloc.rs:6:20
+   |
+LL | fn test_system(_x: system){}
+   |                    ^^^^^^ not found in this scope
+   |
+help: did you mean `std::alloc::System`?
+   |
+LL | fn test_system(_x: std::alloc::System){}
```

Comment:
IMO repeating the whole expression makes the hint look cluttered, and - at least for me - quite hard to follow; what do you say we just print the name of the suggested type?
```
error[E0412]: cannot find type `system` in this scope
  --> $DIR/alloc.rs:6:20
   |
LL | fn test_system(_x: system){}
   |                    ^^^^^^ help: a struct with a similar name exists: `std::alloc::System`
   |
```
For reference, you already get a similar message for full paths:
```rust
struct Foo {
    x: std::collections::HashMaP<(), ()>,
}
```
```
error[E0412]: cannot find type `HashMaP` in module `std::collections`
   --> src/lib.rs:4:26
    |
4   |       x: std::collections::HashMaP<(), ()>,
    |                            ^^^^^^^ help: a struct with a similar name exists: `HashMap`
```

Reply:
The error label on that span is more generally used and is not being defined here.  We will continue to get the "not found in this scope" string without edits to that area of the diagnostics.  The libstd suggestions in this PR are being appended to the error diagnostics that occur further up the type check chain.

I can remove the separate suggestion block.  I intentionally used the verbose form of the suggestion because I find the formatting preferable to the default version.  Here is what I get with the change:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a libstd struct with a similar name `std::collections::HashMap`?: `std::collections::HashMap`

error: aborting due to previous error
```

The suggestion string (here the full struct path) is repeated on the same line following a colon. We could eliminate the suggestion altogether and just format it as a label but IIUC this eliminates the ability for autofixes by tools.

Thoughts?

cc: @euclio with the suggestion string change above

Reply:
The error label on the span ("not found in this scope") is added in a more general check further up the diagnostics chain.  We can't eliminate it without modifying that source.

I intentionally used the verbose version of the suggestion because I tend to like the formatting.  It is set off with a suggestion in a different color and I think that this makes it distinct from the message about what the error is.  I don't have strong opinions about this though.  Here is how it appears with the default suggestion format and a change in wording to address your and @euclio concerns:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a libstd struct with a similar name: `std::collections::HashMap`

error: aborting due to previous error
```

Reply:
Thoughts?

Reply:
Pushed these changes in https://github.com/rust-lang/rust/pull/72988/commits/bf2f78eaab4b9a9044011921bb2f31d18b8bb5ab so that you can see all of the UI test stderr strings with the new format.

Reply:
The new version looks really neat! :smile: 

I'd suggest two adjustments:
- `found libstd structs with a similar name` -> `found libstd structs with similar names` (since `names` refers to those newly-found structs)
- `libstd` -> `std` (I think it's more ubiquitous & newbie-friendly term than `libstd`)

Reply:
Done!

https://github.com/rust-lang/rust/pull/72988/commits/9441340d0be1922b047f82dbcde516ad0d52cac7

Reply:
Here is where we are:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:4:9
  |
4 |     hm: Hashmap<String, ()>,
  |         ^^^^^^^
  |         |
  |         not found in this scope
  |         help: found a std struct with a similar name: `std::collections::HashMap`

error: aborting due to previous error
```

Reply:
wonderful, thanks :-)

Reply:
Thanks for your feedback.  I really appreciate it!